### PR TITLE
Upgrade django

### LINKED ISSUE #903: Upgrade to Django 3.2 LTS
Current Django 2.2 LTS end of support is Apr 2022

Need to upgrade to Django 3.2 LTS (support to Apr 2024)

### PR DESCRIPTION
**Pull request by yujinyuz**
Closes #903

### CODE REVIEW

**Comment by yujinyuz** [comment]
I was able to narrow down the issue that `django-tenants==3.1.0` doesn't break the tests while `django-tenants==3.4.2` does.

I'll need to dig more into this

**Comment by yujinyuz** [comment]
I tried to search for other tests that use `save_model` but it seems that this is an unreliable way to test validators.

Based on django's [documentation](https://docs.djangoproject.com/en/3.2/ref/validators/#how-validators-are-run)

> Note that validators will not be run automatically when you save a model, but if you are using a ModelForm, it will run your validators on any fields that are included in your form. See the ModelForm documentation for information on how model validation interacts with forms.

I'm not sure why it was working before though...

Why do we have this?

https://github.com/bytedeck/bytedeck/blob/9bf9dcd06f756b2fd97ada480887cb632ec0e50a/src/tenant/tests/test_admin.py#L61-L66

**Comment by tylerecouture** [comment]
> 
> Why do we have this?
> 
> https://github.com/bytedeck/bytedeck/blob/9bf9dcd06f756b2fd97ada480887cb632ec0e50a/src/tenant/tests/test_admin.py#L61-L66

I don't remember! It looks like I was trying to test something bigger and couldn't get it to work so most of it is commented out.

**Comment by yujinyuz** [comment]
Okay. I've finally understood why this was getting an error.

TL;DR: It is working for `v3.4.0`. In `v3.4.1`, they allowed schema names with special characters.


 This line in our test case:

https://github.com/bytedeck/bytedeck/blob/9bf9dcd06f756b2fd97ada480887cb632ec0e50a/src/tenant/tests/test_admin.py#L99-L105

It raises a `ValidationError` not because it has an invalid name, but because it has an invalid schema name. 

```
django.core.exceptions.ValidationError: ['Invalid string used for the schema name.']
```

And because `non public` schema name is now allowed in version `3.4.1`, it doesn't raise that error anymore.

Should we comment out the test first and look for a fix later?

**Comment by tylerecouture** [comment]
I think the problem is that when creating a new tenant, we use the name to generate the schema and domain.

**Comment by yujinyuz** [comment]
Btw, we already have tests for `TenantAdminForm` which is what I think should be tested instead of `save_model`.

> I think the problem is that when creating a new tenant, we use the name to generate the schema and domain.


Then what should be addressed should we use to generate the schema name and domain?

**Comment by tylerecouture** [comment]
What about :8000/decks/new?

Will that form still work too? I don't remember if it's covered by tests.

**Comment by yujinyuz** [comment]
@tylerecouture I just checked and it isn't covered by tests.

Apparently, we have `TenantForm` and `TenantAdminForm`

**Comment by tylerecouture** [comment]
What's your suggestion?  Can you just manually test that view and if it's working fine (enter some bad names) carry on?  or can you write some tests for it?

File: requirements-production.txt [hunk 1]
```diff
@@ -1,8 +1,8 @@
-django-tenants>=3.1,<3.2
+django-tenants>=3.4,<3.5
 backports.csv>=1.0.7,<1.1
 beautifulsoup4>=4.8,<4.9
 celery>=4.4,<5.0
-Django>=2.2.9,<2.3
+Django>=3.2,<4
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply requirements-production.txt:5]: @yujinyuz my comment here needs to be addressed in this PR Thanks!

**Review by tylerecouture — CHANGES_REQUESTED** [review]
